lib/ruby_lsp/document.rb

```diff
@@ -190,10 +198,31 @@ def locate(node, char_position, node_types: [])
       # though the class/module node does indeed enclose the target, because it would lead to incorrect behavior
       if closest.is_a?(Prism::ConstantReadNode) || closest.is_a?(Prism::ConstantPathNode)
         last_level = nesting.last
-        nesting.pop if last_level && last_level.constant_path == closest
+
+        if (last_level.is_a?(Prism::ModuleNode) || last_level.is_a?(Prism::ClassNode)) &&
+            last_level.constant_path == closest
+          nesting.pop
+        end
+      end
+
+      complete_nesting = []
+      surrounding_method = T.let(nil, T.nilable(String))
+
+      nesting.each do |node|
+        case node
+        when Prism::ClassNode, Prism::ModuleNode
+          complete_nesting << node.constant_path.slice
+        when Prism::SingletonClassNode
+          complete_nesting << "<Class:#{complete_nesting.last}>"
+        when Prism::DefNode
+          surrounding_method = node.name.to_s
+          next unless node.receiver.is_a?(Prism::SelfNode)
+
+          complete_nesting << "<Class:#{complete_nesting.last}>"
+        end
       end
 
-      NodeContext.new(closest, parent, nesting.map { |n| n.constant_path.location.slice }, call_node)
+      NodeContext.new(closest, parent, complete_nesting, call_node, surrounding_method)
```

Comment:
Can we rename `nesting` to `nesting_nodes`, and `complete_nesting` can simply be `nesting`.